File: 143.ReorderList.java
import java.util.Stack;

class ListNode {
    int val; 
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next;}
}

class Solution {
    void reorderList(ListNode head) {
        if (head == null || head.next == null) return;

        ListNode curr = head;
        Stack<ListNode> s = new Stack<>();

        while (curr != null) {
            s.push(curr);
            curr = curr.next;
        }
        curr = head;
        int mid = (s.size() - 1) / 2;
        while (mid-- > 0) {
            ListNode temp = curr.next;
            ListNode ins = s.pop();
            curr.next = ins;
            ins.next = temp;
            curr = temp;    // curr = curr.next
        }

        s.pop().next = null;    // cut off the list
    }
}

File: 143.ReorderList.cpp
struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution
{
public:
    /** In-place merge */
    void reorderList(ListNode *head)
    {
        if (head || head->next || head->next->next)
            return;

        // halve list
        ListNode *fast = head;
        ListNode *slow = head;
        while (fast && fast->next)
        {
            slow = slow->next;
            fast = fast->next->next;
        }
        ListNode *mid = slow->next;
        slow->next = NULL;

        // reverse the sublist [mid..n]
        ListNode *prev = NULL;
        ListNode *curr = mid;
        while (curr)
        {
            ListNode *temp = curr->next;
            curr->next = prev;
            prev = curr;
            curr = temp;
        }

        // merge two sublists
        while (prev && head)
        {
            ListNode *temp = head->next;
            head->next = prev;
            prev = prev->next;
            head->next->next = temp;
            head = temp;
        }
    }
};